### DISCUSSION TITLE
How to create an adapter for Parse server?

### DISCUSSION THREAD

**Discussion by hinryd** [flutterdata/flutter_data#120] [Q&A] [answered]
Just stumbled across this library seems to solve my data problems, there are a few questions as the documentation is sparse.

1. How do I configure my adapter to recognize the `results` array?
2. How do I change the default `id` field to `objectId`?

Below is an example of the response:
```
{
  "results": [
    {
      "playerName": "Jang Min Chul",
      "updatedAt": "2011-08-19T02:24:17.787Z",
      "cheatMode": false,
      "createdAt": "2011-08-19T02:24:17.787Z",
      "objectId": "A22v5zRAgd",
      "score": 80075
    },
    {
      "playerName": "Sean Plott",
      "updatedAt": "2011-08-21T18:02:52.248Z",
      "cheatMode": false,
      "createdAt": "2011-08-20T02:06:57.931Z",
      "objectId": "Ed1nuqPvcm",
      "score": 73453
    }
  ]
}
```

**Comment by frank06** [comment, marked as answer]
Hi @hinryd , I'm working on the new docs as we speak! I hope to have them up this week.

In Flutter Data we use adapters to override behavior.

Let's assume your model is called `Player`.

(1) We need to pass the deserializer a different root, so we override it:

```dart
mixin PlayerAdapter on RemoteAdapter<Player> {
  @override
  DeserializedData<T, DataModel> deserialize(Object? data,
      {String? key}) {
    // whatever condition makes sense for your data
    if (data['results'] is Iterable) {
      return super.deserialize(data['results'], key: key);
    }
    return super.deserialize(data, key: key);
  }
}
```

If all your models have this behavior (not only players) you can use it like

```dart
mixin ResultsAdapter<T extends DataModel<T>> on RemoteAdapter<T> { }
```

In the model include the adapter and specify a new ID for (2) which really is a `json_serializable` configuration.

```dart
@DataRepository([PlayerAdapter])
@JsonSerializable()
class Player with DataModel<Player> {
  @override
  @JsonKey(name: 'objectId') final String? id;
  final String playerName;
  final int score;
  // ...
  final BelongsTo<Family> family;

  Person({
    this.id,
    required this.playerName,
    this.score,
   // ...
  });
```

Hope this helps!

**Reply by hinryd** [reply]
Thank you Frank! I will definitely recommend this lib to my colleagues!

**Reply by frank06** [reply]
No problem!

FYI I just released the new and complete docs site: https://flutterdata.dev/
